From 4c3468fa6ab0a993f4bc15b503061bbbeedd19e7 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Sat, 4 Jan 2025 13:57:30 +0900
Subject: [PATCH] sd-json: introduce json_dispatch_log_level()

Then, use it in io.systemd.service.SetLogLevel method.

(cherry picked from commit 93081be64b858afea32079a8b32212a36b64de84)

Resolves: RHEL-75774
---
 src/libsystemd/sd-json/json-util.c      | 20 ++++++++++++++++++++
 src/libsystemd/sd-json/json-util.h      |  1 +
 src/shared/varlink-io.systemd.service.c | 12 ++++--------
 3 files changed, 25 insertions(+), 8 deletions(-)

diff --git a/src/libsystemd/sd-json/json-util.c b/src/libsystemd/sd-json/json-util.c
index 67e50e6c51..1b22084d40 100644
--- a/src/libsystemd/sd-json/json-util.c
+++ b/src/libsystemd/sd-json/json-util.c
@@ -327,6 +327,26 @@ int json_dispatch_ifindex(const char *name, sd_json_variant *variant, sd_json_di
         return 0;
 }
 
+int json_dispatch_log_level(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        int *log_level = ASSERT_PTR(userdata), r, t;
+
+        if (sd_json_variant_is_null(variant)) {
+                *log_level = -1;
+                return 0;
+        }
+
+        r = sd_json_dispatch_int(name, variant, flags, &t);
+        if (r < 0)
+                return r;
+
+        /* If SD_JSON_RELAX is set allow a zero interface index, otherwise refuse. */
+        if (LOG_PRI(t) != t)
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not a valid log level.", strna(name));
+
+        *log_level = t;
+        return 0;
+}
+
 int json_variant_new_devnum(sd_json_variant **ret, dev_t devnum) {
         if (devnum == 0)
                 return sd_json_variant_new_null(ret);
diff --git a/src/libsystemd/sd-json/json-util.h b/src/libsystemd/sd-json/json-util.h
index b3b4941dcc..9a2b7b5ff2 100644
--- a/src/libsystemd/sd-json/json-util.h
+++ b/src/libsystemd/sd-json/json-util.h
@@ -119,6 +119,7 @@ int json_dispatch_path(const char *name, sd_json_variant *variant, sd_json_dispa
 int json_dispatch_pidref(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata);
 int json_dispatch_devnum(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata);
 int json_dispatch_ifindex(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata);
+int json_dispatch_log_level(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata);
 int json_dispatch_strv_environment(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata);
 
 static inline int json_variant_unbase64_iovec(sd_json_variant *v, struct iovec *ret) {
diff --git a/src/shared/varlink-io.systemd.service.c b/src/shared/varlink-io.systemd.service.c
index 666778bd41..62cfc9b637 100644
--- a/src/shared/varlink-io.systemd.service.c
+++ b/src/shared/varlink-io.systemd.service.c
@@ -2,7 +2,7 @@
 
 #include <unistd.h>
 
-#include "macro.h"
+#include "json-util.h"
 #include "varlink-io.systemd.service.h"
 
 static SD_VARLINK_DEFINE_METHOD(Ping);
@@ -33,13 +33,12 @@ int varlink_method_ping(sd_varlink *link, sd_json_variant *parameters, sd_varlin
 
 int varlink_method_set_log_level(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
         static const sd_json_dispatch_field dispatch_table[] = {
-                { "level", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_int64, 0, SD_JSON_MANDATORY },
+                { "level", _SD_JSON_VARIANT_TYPE_INVALID, json_dispatch_log_level, 0, SD_JSON_MANDATORY },
                 {}
         };
 
-        int64_t level;
+        int r, level;
         uid_t uid;
-        int r;
 
         assert(link);
         assert(parameters);
@@ -53,9 +52,6 @@ int varlink_method_set_log_level(sd_varlink *link, sd_json_variant *parameters,
         if (r != 0)
                 return r;
 
-        if (LOG_PRI(level) != level)
-                return sd_varlink_error_invalid_parameter(link, parameters);
-
         r = sd_varlink_get_peer_uid(link, &uid);
         if (r < 0)
                 return r;
@@ -63,7 +59,7 @@ int varlink_method_set_log_level(sd_varlink *link, sd_json_variant *parameters,
         if (uid != getuid() && uid != 0)
                 return sd_varlink_error(link, SD_VARLINK_ERROR_PERMISSION_DENIED, parameters);
 
-        log_debug("Received io.systemd.service.SetLogLevel(%" PRIi64 ")", level);
+        log_debug("Received io.systemd.service.SetLogLevel(%i)", level);
 
         log_set_max_level(level);
 
